cli/flox-rust-sdk/src/providers/buildenv.rs

```diff
@@ -0,0 +1,209 @@
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::sync::LazyLock;
+
+use indoc::indoc;
+use serde::Deserialize;
+use thiserror::Error;
+use tracing::debug;
+
+use crate::models::pkgdb::{call_pkgdb, CallPkgDbError, PkgDbError, PKGDB_BIN};
+use crate::utils::CommandExt;
+
+static NIX_BIN: LazyLock<PathBuf> = LazyLock::new(|| {
+    std::env::var("NIX_BIN")
+        .unwrap_or_else(|_| env!("NIX_BIN").to_string())
+        .into()
+});
+
+static BUILDENV_NIX: LazyLock<PathBuf> = LazyLock::new(|| {
+    std::env::var("FLOX_BUILDENV_NIX")
+        .unwrap_or_else(|_| env!("FLOX_BUILDENV_NIX").to_string())
+        .into()
+});
+
+#[derive(Debug, Error)]
+pub enum BuildEnvError {
+    /// An error that occurred while realising the packages in the lockfile.
+    /// Those are Nix errors pkgdb forwards to us as well as detection of
+    /// disallowed packages.
+    #[error("Failed to realise packages in lockfile")]
+    Realise(#[source] PkgDbError),
+
+    /// Other errors arising from calling pkgdb and interpreting its output.
+    #[error(transparent)]
+    CallPkgDb(CallPkgDbError),
+
+    /// An error that occurred while composing the environment.
+    /// I.e. `nix build` returned with a non-zero exit code.
+    /// The error message is the stderr of the `nix build` command.
+    // TODO: this requires to capture the stderr of the `nix build` command
+    // or essentially "tee" it if we also want to foward the logs to the user.
+    // At the moment the "interesting" logs
+    // are emitted by the `pkgdb realise` portion of the build.
+    // So in the interest of initial simplicity
+    // we can defer forwarding the nix build logs and capture output with [Command::output].
+    #[error("Failed to compose environment: {0}")]
```

Comment:
The "compose" terminology isn't super obvious and I don't think it will mean anything to an end user that sees this message. Maybe "construct" instead?